File: src/main/java/com/example/demo/service/UbigeoIneiService.java
package com.example.demo.service;

import com.example.demo.dto.UbigeoCoberturaResponse;
import com.example.demo.dto.UbigeoResponse;
import java.util.List;

public interface UbigeoIneiService {
    List<String> listarDepartamentos();

    List<String> listarProvinciasPorDepartamento(String departamento);

    List<String> listarDistritosPorDepartamentoYProvincia(String departamento, String provincia);

    UbigeoResponse consultarUbigeo(String departamento, String provincia, String distrito);

    UbigeoCoberturaResponse consultarCoberturaPorUbigeo(String ubigeo);

    UbigeoResponse consultarDetallePorUbigeo(String ubigeo);
}

File: frontend/src/app/core/services/ubigeo-inei.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { API_ENDPOINTS } from '../constants/api-endpoints';
import { UbigeoCoberturaResponse, UbigeoResponse } from '../models/ubigeo-inei.models';
import { ApiService } from './api.service';

@Injectable({ providedIn: 'root' })
export class UbigeoIneiService {
  private readonly api = inject(ApiService);
  private readonly basePath = API_ENDPOINTS.v1.ubigeos;

  listarDepartamentos(): Observable<string[]> {
    return this.api.get<string[]>(`${this.basePath}/departamentos`);
  }

  listarProvincias(departamento: string): Observable<string[]> {
    return this.api.get<string[]>(`${this.basePath}/provincias`, { departamento });
  }

  listarDistritos(departamento: string, provincia: string): Observable<string[]> {
    return this.api.get<string[]>(`${this.basePath}/distritos`, { departamento, provincia });
  }

  consultarCodigoUbigeo(departamento: string, provincia: string, distrito: string): Observable<UbigeoResponse> {
    return this.api.get<UbigeoResponse>(`${this.basePath}/codigo`, { departamento, provincia, distrito });
  }

  consultarCoberturaPorUbigeo(ubigeo: string): Observable<UbigeoCoberturaResponse> {
    return this.api.get<UbigeoCoberturaResponse>(`${this.basePath}/${ubigeo}/cobertura`);
  }
}
